beluga/include/beluga/sensor/likelihood_3d_field_model.hpp

```diff
@@ -0,0 +1,167 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_SENSOR_LIKELIHOOD_3D_FIELD_MODEL_HPP
+#define BELUGA_SENSOR_LIKELIHOOD_3D_FIELD_MODEL_HPP
+
+#include <algorithm>
+#include <cmath>
+#include <random>
+#include <vector>
+
+#include <openvdb/openvdb.h>
+#include <openvdb/tools/VolumeToSpheres.h>
+
+#include <Eigen/Core>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/transform.hpp>
+#include <sophus/se3.hpp>
+#include <sophus/so3.hpp>
+
+/**
+ * \file
+ * \brief Implementation of a likelihood field sensor model for 3D Lidars.
+ */
+
+namespace beluga {
+
+/// Parameters used to construct a Likelihood3DFieldModel instance.
+/**
+ * See Probabilistic Robotics \cite thrun2005probabilistic Chapter 6.4, particularly Table 6.3.
+ */
+struct Likelihood3DFieldModelParam {
+  /// Voxel size in meters.
+  double voxel_size = 0.07;
+  /// Maximum distance to obstacle.
+  /**
+   * When creating a distance map, if the distance to an obstacle is higher than the value specified here,
+   * then this value will be used.
+   */
+  double max_obstacle_distance = 100.0;
+  /// Maximum range of a laser ray.
+  double max_laser_distance = 2.0;
+  /// Weight used to combine the probability of hitting an obstacle.
+  double z_hit = 0.5;
+  /// Weight used to combine the probability of random noise in perception.
+  double z_random = 0.5;
+  /// Standard deviation of a gaussian centered arounds obstacles.
+  /**
+   * Used to calculate the probability of the obstacle being hit.
+   */
+  double sigma_hit = 0.2;
+};
+
+/// Likelihood field sensor model for range finders.
+/**
+ * This model relies on a pre-computed likelihood map of the environment.
+ * It is less computationally intensive than the beluga::BeamSensorModel
+ * because no ray-tracing is required, and it can also provide better
+ * performance in environments with non-smooth occupation maps. See
+ * Probabilistic Robotics \cite thrun2005probabilistic, Chapter 6.4,
+ * for further reference.
+ *
+ * \note This class satisfies \ref SensorModelPage.
+ *
+ * \tparam OpenVDB grid type.
+ */
+template <typename GridT, typename U>
```

Comment:
@pvela2017 consider using meaningful template parameter names (e.g. `PointCloud` instead of `U`)

Reply:
Changed in 015f1b79441f103c2e55d662d87fa4c144e0ef2c